fiona/ogrext.pyx

```diff
@@ -1287,6 +1295,37 @@ def _remove(path, driver=None):
         raise RuntimeError("Failed to remove data source {}".format(path))
 
 
+def _remove_layer(path, layer, driver=None):
+    cdef void *cogr_ds
+    cdef int layer_index
+    
+    if isinstance(layer, integer_types):
+        layer_index = layer
+        layer_str = str(layer_index)
+    else:
+        layer_names = _listlayers(path)
+        try:
+            layer_index = layer_names.index(layer)
+        except ValueError:
+            raise ValueError("Layer \"{}\" does not exist in datasource: {}".format(layer, path))
+        layer_str = '"{}"'.format(layer)
+    
+    if layer_index < 0:
+        layer_names = _listlayers(path)
+        layer_index = len(layer_names) + layer_index
+    
+    cogr_ds = gdal_open_vector(path.encode("utf-8"), 1, None, {})
+    if cogr_ds == NULL:
+        raise IOError("Failed to open data source: {}".format(path))
+
+    result = OGR_DS_DeleteLayer(cogr_ds, layer_index)
```

Comment:
Can we use `exc_wrap_int` here?

Reply:
There are a few places we could apply this (anywhere `OGRERR_*` are used). Perhaps better to do this systematically as a separate PR?